syntax = "proto3";

service TaskScheduler {
    rpc RegisterUser(RegisterRequest) returns (UserResponse);
    rpc LoginUser(LoginRequest) returns (LoginResponse);
    rpc ScheduleTask(TaskRequest) returns (TaskResponse);
    rpc GetTaskStatus(TaskStatusRequest) returns (TaskStatusResponse);
    rpc ListTasks(ListTasksRequest) returns (ListTasksResponse);
    rpc ListHistory(ListHistoryRequest) returns (ListHistoryResponse);
}

message RegisterRequest {
    string email = 1;
    string name = 2;
    string password = 3;
}

message LoginRequest {
    string email = 1;
    string name = 2;
    string password = 3;
}

message LoginResponse {
    string token = 1;
    bool is_admin = 2;
    string name = 3; 
}
message UserResponse {
    string user_id = 1;
    string message = 2;
}

message TaskRequest {
    string name = 1;
    string description = 2;
    string schedule_time = 3;
}

message TaskResponse {
    string task_id = 1;
    string status = 2;
    string worker_id = 3;
}

message TaskStatusRequest {
    string task_id = 1;
}

message TaskStatusResponse {
    string task_id = 1;
    string status = 2;
    string details = 3;
    string worker_id = 4;
}

message ListTasksRequest {}

message TaskInfo {
    string task_id = 1;
    string name = 2;
    string description = 3;
    string schedule_time = 4;
    string status = 5;
    string worker_id = 6;
    string completion_time = 7;
}

message ListTasksResponse {
    repeated TaskInfo tasks = 1;
}

message ListHistoryRequest {}

message HistoryEntry {
    string task_id = 1;
    string name = 2;
    string description = 3;
    string worker_id = 4;
    string completion_time = 5;
}

message ListHistoryResponse {
    repeated HistoryEntry history = 1;
}
